From 59c5e055af21f22c8dd0308a2f42834655cd8e6f Mon Sep 17 00:00:00 2001
From: ralves <myeyeismadeofglass@gmail.com>
Date: Tue, 18 Jan 2022 15:10:20 -0300
Subject: [PATCH 3/4] Add BUILDROOT_USERSPACE build option

---
 CMakeLists.txt | 16 ++++++++++++++++
 1 file changed, 16 insertions(+)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 2b9b321..537a319 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -47,6 +47,8 @@ if(CMAKE_SYSTEM_NAME MATCHES "Linux")
         MESSAGE("TOOLCHAIN = GCC")
     elseif(TOOLCHAIN STREQUAL "BUILDROOT")
         MESSAGE("TOOLCHAIN = BUILDROOT")
+    elseif(TOOLCHAIN STREQUAL "BUILDROOT_USERSPACE")
+        MESSAGE("TOOLCHAIN = BUILDROOT_USERSPACE")
     elseif(TOOLCHAIN STREQUAL "CLANG")
         MESSAGE("TOOLCHAIN = CLANG")
     elseif(TOOLCHAIN STREQUAL "CBMC")
@@ -164,6 +166,20 @@ if(CMAKE_SYSTEM_NAME MATCHES "Linux")
 
         SET(CMAKE_C_LINK_EXECUTABLE "<CMAKE_LINKER> <LINK_FLAGS> <OBJECTS> -o <TARGET> -Wl,--start-group <LINK_LIBRARIES> -Wl,--end-group")
 
+    elseif(TOOLCHAIN STREQUAL "BUILDROOT_USERSPACE")
+        SET(CMAKE_C_COMPILER x86_64-buildroot-linux-uclibc-gcc)
+        SET(CMAKE_C_FLAGS "-g -fshort-wchar -fno-strict-aliasing -Wall         -Wno-array-bounds -ffunction-sections -fdata-sections -fno-common -maccumulate-outgoing-args -mno-red-zone -Wno-address       -fno-asynchronous-unwind-tables                   -Wno-switch -Wno-maybe-uninitialized -Wno-uninitialized                                    -Wno-nonnull-compare ")
+        SET(MBEDTLS_FLAGS "")
+        SET(OPENSSL_FLAGS "-include base.h -Wno-error=maybe-uninitialized -Wno-error=format -Wno-format -Wno-error=unused-but-set-variable")
+        SET(CMOCKA_FLAGS "-std=gnu99 -Wpedantic -Wall -Wshadow -Wmissing-prototypes -Wcast-align -Werror=address -Wstrict-prototypes -Werror=strict-prototypes -Wwrite-strings -Werror=write-strings -Werror-implicit-function-declaration -Wpointer-arith -Werror=pointer-arith -Wdeclaration-after-statement -Werror=declaration-after-statement -Wreturn-type -Werror=return-type -Wuninitialized -Werror=uninitialized -Werror=strict-overflow -Wstrict-overflow=2 -Wno-format-zero-length -Wmissing-field-initializers -Wformat-security -Werror=format-security -fno-common -Wformat -fno-common ") # -fstack-protector-strong
+
+        SET(CMAKE_AR x86_64-buildroot-linux-uclibc-gcc-ar)
+
+        SET(CMAKE_LINKER x86_64-buildroot-linux-uclibc-gcc)
+        SET(CMAKE_EXE_LINKER_FLAGS "      -Wno-error         --coverage -lgcov -fprofile-arcs -ftest-coverage" )
+
+        SET(CMAKE_C_LINK_EXECUTABLE "<CMAKE_LINKER> <LINK_FLAGS> <OBJECTS> -o <TARGET> -Wl,--start-group <LINK_LIBRARIES> -Wl,--end-group")
+
     elseif(TOOLCHAIN STREQUAL "ARM_GCC")
         SET(CMAKE_C_COMPILER arm-linux-gnueabi-gcc)
         SET(CMAKE_C_FLAGS "-g -fshort-wchar -fno-strict-aliasing -Wall -Werror -Wno-array-bounds -ffunction-sections -fdata-sections -fno-common -Wno-address -fpie -fno-asynchronous-unwind-tables -flto -DUSING_LTO -Wno-switch -Wno-maybe-uninitialized -Wno-uninitialized  -Wno-builtin-declaration-mismatch -Wno-nonnull-compare --coverage -fprofile-arcs -ftest-coverage")
-- 
2.17.5

